### PR TITLE
Two build fixes

### PR DESCRIPTION
**Pull request by stanislav-brabec**

### CODE REVIEW

**Comment by stanislav-brabec** [comment]
Closing - one check failed.

**Comment by jimklimov** [comment]
Note you are not required to close a reasonable PR just because a build/check failed. You can push further commits to the same branch you coded your changes in, and they will become automatically parts of the PR and re-tested - until it's green.

It might make sense to peer-review unrelated changes as different PRs though ;)

In your developer fork, you can `git create -b` more branches to post the separate improvements as separate PRs, and for this particular case, after branching off the current state, edit your `master` branch with e.g. `git rebase -i HEAD~2` to remove the failing HTML documentation commit from this branch and then force-push it back to your github fork. This should leave just one (working) commit that can be merged as a PR for systemd improvement :)

As for HTML improvement, according to https://travis-ci.org/networkupstools/nut/jobs/223607016 there is some access issue - not sure OTOH what it can be, perhaps the read-onlyness of some paths during `make distcheck`, or that the JPG files are not `EXTRA_DIST`ed etc. - this needs a bit more investigation and fix (on your side? ;) ): 

````
2927 /usr/bin/a2x  --attribute icons --xsltproc-opts "--nonet" --xsltproc-opts "--stringparam nut.localdate \"`TZ=UTC date +%Y-%m-%d`\"" --xsltproc-opts "--stringparam nut.localtime \"`TZ=UTC date +%H:%M:%S`\"" --xsltproc-opts "--stringparam nut.nutversion \"2.7.4.1\"" --attribute iconsdir=../../docs/images --attribute=badges --attribute=external_title --attribute tree_version=2.7 -a toc -a numbered --destination-dir=. --attribute=xhtml11_format --format=xhtml --xsl-file=../../docs/xhtml.xsl ../../docs/cables.txt
2928 Traceback (most recent call last):
2929  File "/usr/bin/a2x", line 900, in <module>
2930    a2x.execute()
2931  File "/usr/bin/a2x", line 352, in execute
2932    self.__getattribute__('to_'+self.format)()
2933  File "/usr/bin/a2x", line 605, in to_xhtml
2934    self.copy_resources(xhtml_file, src_dir, self.destination_dir)
2935  File "/usr/bin/a2x", line 581, in copy_resources
2936    shell_copy(src, dst)
2937  File "/usr/bin/a2x", line 161, in shell_copy
2938    shutil.copy(src, dst)
2939  File "/usr/lib/python2.7/shutil.py", line 117, in copy
2940    copyfile(src, dst)
2941  File "/usr/lib/python2.7/shutil.py", line 83, in copyfile
2942    with open(dst, 'wb') as fdst:
2943 IOError: [Errno 13] Permission denied: u'/home/travis/build/networkupstools/nut/nut-2.7.4.1/_build/docs/images/cables/73-0724.png'
2944 make[2]: *** [cables.html] Error 1
2945 make[2]: Leaving directory `/home/travis/build/networkupstools/nut/nut-2.7.4.1/_build/docs'
````

Technically, if someone in the team comes up with a fix for the HTML issue (or someone else posts a PR expanding on your fix in https://github.com/networkupstools/nut/pull/425/commits/22d1ec162956c7ba6450012a10d5bf9c87be3f12), your PR can be merged even based on the "red" code, followed shortly by the fix, so upstream/master stays green. But ATM I am not sure I can dedicate time to this on short notice :)

**Comment by stanislav-brabec** [comment]
I just analyzed the failure. It can happen only if builddir != srcdir. But it does not happen every time. Sometimes it even happens for other doc files than cables.

Directory "images" is copied to the build directory with read-only permission (probably by a2x). Normally it is no problem, but sometimes a2x fails on that. I did not yet found a source of the problem.

"make" fails
"make ; make" succeeds
"make clean ; make" fails
"rm -rf images ; make" succeeds in some cases

**Comment by jimklimov** [comment]
Well, "builddir != srcdir" is the domain of distcheck, which on purpose does out-of-tree builds to make sure recipes are correct - or expose presence of errors, in this case. The "dist" part copies current sources to the other directory, and the "check" one tries to build that copy in out-of-tree mode - making sure in particular that all files are in place in the copy, so that "dist" directory can be tarballed as the source distributive. Also by design, the "disted" sources are kept read-only as they can be available from only RO media like CDROM or an un-pollutable network share - and builds are done in local FS of the worker.

For simple 1:1 rules it is usually `make` and its `VPATH` processing that does the magic regarding a full path to an object (input and output of a recipe), especially if there are no conflicting matching choices to pick from :)

Sometimes it does seem to require workarounds, such as requesting the explicit src/bld dirs in the recipe. Usually I try to keep those reserved to recipes that deal with maintenance of the codebase itself - e.g. for interactive spellchecker recipes and using the dictionary file in the actual workspace, not in some dist-dir. 

Also note that depending on `configure --with-doc=...` options, NUT build procedure can by default require presence and "copy-ability" of documentation sources (text files, not sure OTOH if images are covered) and ability to build at least manpages from them, *or* it can take explicitly requested shortcuts to not build (skip) documentation including in `distcheck` to speed up the test builds. It is marginally possible that something in this area bites you :\

In a different project, such customized dist things were handled by either `dist-hook` definitions or via Make dependencies, to handle both in-tree and out-of-tree builds (where tools like `a2x` in this case, or selftest programs in that case, require an exact path relative to the binary or current working directory) using a pattern like this for an `src/required.file` as example:

````
EXTRA_DIST += \
        src/required.file

$(abs_builddir)/src/required.file: $(abs_srcdir)/src/required.file
        @if [ "$@" != "$<" ]; then cp -pf "$<" "$@" ; fi

# For in-tree builds, the file is just there (src==bld); for out-of-tree it is copied above
check-local: $(abs_builddir)/src/required.file

clean-local: clean-local-check
.PHONY: clean-local-check
clean-local-check:
        @if [ "$(abs_builddir)" != "$(abs_srcdir)" ]; then rm -f $(abs_builddir)/src/required.file ; fi
````

Feel free to adapt, if deemed needed for images in this case...

**Comment by jimklimov** [comment]
Regarding HTML improvement, I happened to strike a similar issue just building a distcheck with HTML enabled (this takes inordinate amounts of time on my box so usually I disable that). The equivalent command (with verbosity added) also fails in shell:

````
jim@jimoi:~/nut/nut-2.7.4.1/_build/sub/docs$ /usr/bin/a2x -v --attribute icons --xsltproc-opts "--nonet" --xsltproc-opts "--stringparam nut.localdate \"`TZ=UTC date +%Y-%m-%d`\"" --xsltproc-opts "--stringparam nut.localtime \"`TZ=UTC date +%H:%M:%S`\"" --xsltproc-opts "--stringparam nut.nutversion \"2.7.4.1\"" --attribute iconsdir=../../../docs/images --attribute=badges --attribute=external_title --attribute tree_version=2.7 -a toc -a numbered --destination-dir=. --attribute=xhtml11_format --format=xhtml --xsl-file=../../../docs/xhtml.xsl ../../../docs/developer-guide.txt

...

asciidoc: include: /home/jim/nut/nut-2.7.4.1/lib/README
asciidoc: developer-guide.txt: line 172: reading: /home/jim/nut/nut-2.7.4.1/lib/README

a2x: executing: "xmllint" --nonet --noout --valid "/home/jim/nut/nut-2.7.4.1/_build/sub/docs/developer-guide.xml"


a2x: chdir /home/jim/nut/nut-2.7.4.1/_build/sub/docs
a2x: executing: "xsltproc" --nonet --stringparam nut.localdate "2017-04-21" --stringparam nut.localtime "11:54:27" --stringparam nut.nutversion "2.7.4.1" --stringparam callout.graphics 0 --stringparam navig.graphics 0 --stringparam admon.textlabel 1 --stringparam admon.graphics 0  --output "/home/jim/nut/nut-2.7.4.1/_build/sub/docs/developer-guide.html" "/home/jim/nut/nut-2.7.4.1/docs/xhtml.xsl" "/home/jim/nut/nut-2.7.4.1/_build/sub/docs/developer-guide.xml"


a2x: chdir /home/jim/nut/nut-2.7.4.1/_build/sub/docs
a2x: finding resources in: /home/jim/nut/nut-2.7.4.1/_build/sub/docs/developer-guide.html
a2x: finding resources in: /home/jim/nut/nut-2.7.4.1/_build/sub/docs/developer-guide.html
a2x: copying "/etc/asciidoc/stylesheets/docbook-xsl.css" to "/home/jim/nut/nut-2.7.4.1/_build/sub/docs/docbook-xsl.css"
Traceback (most recent call last):
  File "/usr/bin/a2x", line 958, in <module>
    a2x.execute()
  File "/usr/bin/a2x", line 378, in execute
    self.__getattribute__('to_'+self.format)()
  File "/usr/bin/a2x", line 656, in to_xhtml
    self.copy_resources(xhtml_file, src_dir, self.destination_dir)
  File "/usr/bin/a2x", line 623, in copy_resources
    shell_copy(src, dst)
  File "/usr/bin/a2x", line 176, in shell_copy
    shutil.copy(src, dst)
  File "/usr/lib/python2.7/shutil.py", line 119, in copy
    copyfile(src, dst)
  File "/usr/lib/python2.7/shutil.py", line 83, in copyfile
    with open(dst, 'wb') as fdst:
IOError: [Errno 13] Permission denied: u'/home/jim/nut/nut-2.7.4.1/_build/sub/docs/docbook-xsl.css'


$ ls -la /etc/asciidoc/stylesheets/docbook-xsl.css /home/jim/nut/nut-2.7.4.1/_build/sub/docs/docbook-xsl.css
-r--r--r-- 1 root bin 5738 Oct 17  2014 /etc/asciidoc/stylesheets/docbook-xsl.css
-r--r--r-- 1 jim  jim 5738 Apr 21 13:23 /home/jim/nut/nut-2.7.4.1/_build/sub/docs/docbook-xsl.css
````

This file is not present in my original sources as well as dist'ed ones - so it was probably copied from the OS while building a previous HTML doc and now can not be copied to build a next one... cool :\

And first of all, no idea WHY it is being copied in the first place... this particular one seems like an `a2x` bug, or an older version on this OS, or wrong CLI args...

File: docs/Makefile.am [hunk 1]
```diff
@@ -50,11 +50,13 @@ EXTRA_DIST += $(ALL_TXT_SRC) $(SHARED_DEPS) $(IMAGE_FILES) \
 ASCIIDOC_HTML_SINGLE = user-manual.html \
 	developer-guide.html \
 	packager-guide.html \
+	cables.html \
 	FAQ.html
 
 ASCIIDOC_HTML_CHUNKED = user-manual.chunked \
 	developer-guide.chunked \
 	packager-guide.chunked \
+	cables.html \
```
Comment (jimklimov) [inline docs/Makefile.am]: should be `cables.chunked` here ;)

Comment (stanislav-brabec) [inline docs/Makefile.am]: I think that not. `FAQ` and `cables` are short files, and it makes no sense to make chunked version.

**Comment by jimklimov** [comment]
Thanks for bringing up those issues. They seemed simple, but hope after a day of detective work I nailed them for real :-) See PRs #426 and #428

**Comment by stanislav-brabec** [comment]
The file is copied by a2x itself during the build. The source of the problem seems to be a2x itself. It copies files (from srcdir or its own files) with r/o permission, without checking whether the file already exists and without unlocking previous copy.

I think that a work-around could be `chmod +w` after each command.

The bug is independent on my fix. It just makes it more visible, as it makes impossible to use the same image twice (either by two builds of one doc file, or by one build of two doc files).

**Comment by jimklimov** [comment]
Yes, I came to the same conclusion. In my PR, I `chmod` the working directory before starting a document processing run, but overall it should be the same (at least for sequential builds).

I can also imagine how this explodes for parallel builds. And even if it is a bug fixed in `a2x` say tomorrow, we still have old build systems to support with the issue present for years to come...

**Comment by stanislav-brabec** [comment]
I finally found a complete explanation of the failure:

1. `make distcheck` makes a dist tarball, unpacks it, and calls `chmod -w -R` on the whole unpacked contents.
2. `make distcheck` then creates a build dir.
3. `a2x` detects build in builddir, calls `shutil.copy()`.
4. `shutil.copy()` calls `shutil.copymode()`. It sets mode `0444` on the target file.
5. Next call of `a2x` attempts to copy the same file to a read only location. It fails.

**Comment by jimklimov** [comment]
As I noted in another of this family of issues, this also happened to me with a CSS file that a2x itself copies from the OS. And the OS packages the file as read-only. So this causes effects of 1+2 in your analysis, then the rest of the issue seems the same :)

**Comment by jimklimov** [comment]
BTW, @stanislav-brabec : if you agree that my PRs cover your original intents (and do work for you), feel free to close this one as superseded :) At least, those ended up green on Travis :)

**Comment by stanislav-brabec** [comment]
Superseded by https://github.com/networkupstools/nut/pull/426

**Comment by jimklimov** [comment]
Then omit the file from the list ;)

**Pull request closed without merging**